crates/ruff/src/rules/perflint/rules/manual_dict_comprehension.rs

```diff
@@ -0,0 +1,112 @@
+use rustpython_parser::ast::{self, Expr, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for `for` loops that can be replaced by a dictionary comprehension.
+///
+/// ## Why is this bad?
+/// When creating a dictionary in a for-loop, prefer a dictionary
+/// comprehension. Comprehensions are more readable and more performant.
+///
+/// Using the below as an example, the dictionary comprehension is ~10% faster
+/// on Python 3.11, and ~25% faster on Python 3.10.
+///
+/// Note that, as with all `perflint` rules, this is only intended as a
+/// micro-optimization, and will have a negligible impact on performance in
+/// most cases.
+///
+/// ## Example
+/// ```python
+/// pairs = (("a", 1), ("b", 2))
+/// result = {}
+/// for x, y in pairs:
+///     if y % 2:
+///         result[x] = y
+/// ```
+///
+/// Use instead:
+/// ```python
+/// pairs = (("a", 1), ("b", 2))
+/// result = {x: y for x, y in pairs if y % 2}
+/// ```
+#[violation]
+pub struct ManualDictComprehension;
+
+impl Violation for ManualDictComprehension {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use a dictionary comprehension instead of a for-loop")
+    }
+}
+
+/// PERF404
+pub(crate) fn manual_dict_comprehension(checker: &mut Checker, target: &Expr, body: &[Stmt]) {
+    let [stmt] = body else {
+        return;
+    };
+
+    // For a dictionary comprehension to be appropriate, the loop needs both an index
+    // and a value, so the target must be a tuple.
+    let Expr::Tuple(ast::ExprTuple { elts, .. }) = target else {
+        return;
+    };
+
+    let names = elts
+        .iter()
+        .filter_map(|elt| {
+            if let Expr::Name(ast::ExprName { id, .. }) = elt {
+                Some(id.as_str())
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if names.is_empty() {
+        return;
+    }
+
+    if let Stmt::If(ast::StmtIf { body: if_body, .. }) = stmt {
```

Comment:
Similar to `PERF401`, we should ignore this rule if the dictionary assignment variable is being used inside the conditional test: https://github.com/astral-sh/ruff/blob/30bec3fcfa01c044ea232aae9d3c69b59f422d30/crates/ruff/src/rules/perflint/rules/manual_list_comprehension.rs#L129-L152

Reply:
Good call, I'll make a helper func out of it.

Reply:
I think this is deviating from upstream plugin, and there's no autofix yet, but I do agree it's a good check to have. Added some extra tests (and some of previous are now invalid) and made a helper func out of the check. Got a Clippy warning about borrowing a Box there though I didn't have time to dig into, so let me know if that's not OK to allow and what I could do better there.